### ISSUE TITLE
Fixpoint/Function cannot guess decreasing argument with let bindings

### ISSUE THREAD

**Issue by robblanco** [coq/coq#8226]
#### Version

8.8.1

#### Operating system

Ubuntu 18.04.1

#### Description of the problem

The Coq project in [this folder](https://github.com/AdaCore/sparkformal/tree/master/spark2014_semantics/src) (current commit AdaCore/sparkformal@d1623ff989b2eed35bc494e8f32368090ba08054) fails to build under Coq 8.8.

Using Coq 8.7, the following warning is emitted, but compilation succeeds.

```
File "./rt_gen_impl.v", line 185, characters 0-398:
Warning: Cannot build inversion information
[funind-cannot-build-inversion,funind]
```

After upgrading to Coq 8.8, this becomes an error.

```
File "./rt_gen_impl.v", line 185, characters 0-398:
Error: Cannot guess decreasing argument of fix.
```

The problematic `Function` definition `toObjDeclRTImpl` has a non-recursive body, which is rejected by `Fixpoint` with the same error and accepted by `Definition`. Inlining the `let` binding as shown in robblanco/sparkformal@b5654586606c7f6b9947de8518add7fdaa2d904a serves as a workaround for Coq 8.8 at the moment.

**Comment by siddharthist** [comment]
https://github.com/AdaCore/sparkformal/blob/d1623ff989b2eed35bc494e8f32368090ba08054/spark2014_semantics/src/rt_gen_impl.v#L185

**Comment by Matafou** [comment]
Strange regression. I don't think the code of Function has been modified significantly since 8.7. @mattam82 do you have a clue?

**Comment by ppedrot** [comment]
This might be due to the fact that fixpoints over non-recursive datatypes have been forbidden since.

**Comment by Matafou** [comment]
Mmh ok so Functin should build a simple constant definition here, not a fixpoint.

**Comment by Matafou** [comment]
Ho but Fixpoint should also fail more gracefully.

**Comment by Matafou** [comment]
And here is a minimal example:
```
Record t : Set := mk_t
  { xxx : nat;
    yyy : bool }.

Fixpoint fff (n m:t) : bool :=
match n with
| mk_t x y => 
  let res := mk_t 0 y in
  y
end.
```
Note that adding { struct n } or { struct m } makes fixpoint fail gracefully. The problem comes from the heuristic to chose automatically the decreasing argument. It must be updated to not chose an argument of a non-recursive inductive type.

@forestjulien there must be something to do on Function side to trigger a simple Definition instead of a Fixpoint in that case.

**Comment by Matafou** [comment]
I renamed this bug to split it into two separate bugs: the current one is about a misleading error message: when there is no recursive call in a fixpoint definition the error message should not be "cannot guess recursive argument" but "use Definition instead".

The other bug is to make Function react by calling "Definition"'s backend intead of fixpoint's.

**Comment by SkySkimmer** [comment]
The Fixpoint restriction should really be about if it's a primitive record IMO

**Comment by herbelin** [comment]
Related are:
- PR #407: forbids Fixpoint on types internally stamped as BiFinite (that is - primitive or not - Record and Variant) motivated by the fact that encapsulating a projection of a primitive Record behind a fixpoint leads to subject reduction failure (I'm however not convinced that no other solution is possible like eta-expanding on the fly such an encapsulated projection). This is why @skyskimmer says the flag should be restricted a minima to primitive record.
- Issue #7262 which basically raises the same question: should we align the internal (co)recursive/non-recursive stamp to reflect the reality?

Loosely related:
- PR #6951 about not inserting a `fix` in Scheme Equality when the type is BiFinite (actually, it could be so also when the type is actually non-recursive rather than stamped as recursive).

Unless some users are using `Fixpoint` to artificially block some reduction (???), what would we indeed lose ortherwise at taking the reality into account rather than the Variant/Record/Inductive keyword, as @Alizter is proposing?

It is said in #407 that there is also an issue with parametricity failing with `Fixpoint` (on non-recursive `Prop`?) but I did not get well.

**Comment by herbelin** [comment]
I agree with @skyskimmer that we should a minima modify #407 so that it is only about negative records, i.e. about types without `match`, since somehow, a `fix` over a finite type is like a `match` over this type (that is `fix f x := body` behaves wrt `body` like `fun x => match x with {|p:=x|}|} => body[x:={|p:=x|}] end` does. So, if we don't want `match` for negative types, we don't want `fix` either.

As for non-recursive positive types, I'm arriving to the conclusion that allowing fixpoints is otherwise both hurtless and useless.

Hurtless if we assume that we eventually want eta for all non-recursive types, including the positive ones (`match`-based), since under this assumption and eta for dependent product, `fix f x := body` and `fun x => body` are convertible.

Useless because if the user purpose is to deliberately block the substitution into `body`, we could equally block the substitution by encapulating the body under a `fun x => match x with C y => body[x:=C y] end`.

[I however skip the question of how to express eta for the positive types with indices as the design space seems more open (at least to my knowledge).]

**Comment by herbelin** [comment]
In any case, we need a way to know if a type is really recursive or not (for instance for producing schemes). Do we want to use for that the `mind_finite` field of inductive types or do we prefer instead to keep the `mind_finite` field to remind the intention of the users when they chose between `Inductive`/`CoInductive` or `Record/Variant`?

Personally, I would lean towards keeping both informations (somehow with the idea that Inductive = initial algebra and CoInductive = terminal co-algebra even if degenerate ones, while Record = limit and Variant = colimit).

Then, the other questions I see are:
- do we want to allow `fix`/`cofix` on non-recursive respectively positive/negative types; it does not hurt but the usefulness is unclear? (I think I would personally lean for a no.)
- if yes, do we want schemes on non-recursive types introduced by `Inductive`/`CoInductive` to insert a dummy `fix`/`cofix`?
- incidentally, do we want `induction`/`elim` to conventionally use the `ind` scheme even on non-recursive types, or to fallback on `destruct`/`case` on non-recursive types?
